### PR TITLE
Streaming Guides 

### PR DESCRIPTION
**Pull request by freddyaboulton**
## Description

Streaming guides that cover:
* Streaming Audio and Image inputs. Explains the new parameters `time_limit` and `stream_every`. 
* In-depth streaming audio with transformers demo. This is an audio-in and audio-out demo. 
* Makes some changes to the image component to allow for unified image-to-image streaming demo

Streaming with Transformers Demo: https://huggingface.co/spaces/gradio/magic-8-ball

## 🎯 PRs Should Target Issues

Before your create a PR, please check to see if there is [an existing issue](https://github.com/gradio-app/gradio/issues) for this change. If not, please create an issue before you create this PR, unless the fix is very small. 

Not adhering to this guideline will result in the PR being closed. 

## Tests

1. PRs will only be merged if tests pass on CI. To run the tests locally, please set up [your Gradio environment locally](https://github.com/gradio-app/gradio/blob/main/CONTRIBUTING.md) and run the tests: `bash scripts/run_all_tests.sh`

2. You may need to run the linters: `bash scripts/format_backend.sh` and `bash scripts/format_frontend.sh`

### CODE REVIEW

**Comment by freddyaboulton** [comment]
The guides are good for review but something seems to be up with the e2e tests

File: js/spa/test/cancel_events.spec.ts [hunk 1]
```diff
@@ -7,20 +7,19 @@ test("when using an iterative function the UI should update over time as iterati
 	const start_button = await page.locator("button", {
 		hasText: /Start Iterating/
 	});
-	const textbox = await page.getByLabel("Iterative Output");
```
Comment (abidlabs) [inline js/spa/test/cancel_events.spec.ts:10]: Why this change?

File: demo/magic_8_ball/requirements.txt [hunk 2]
```diff
@@ -0,0 +1,3 @@
+https://gradio-builds.s3.amazonaws.com/bed454c3d22cfacedc047eb3b0ba987b485ac3fd/gradio-4.40.0-py3-none-any.whl
```
Comment (abidlabs) [inline demo/magic_8_ball/requirements.txt:1]: Remove hardcoded Gradio? Or do we need this for now?

File: demo/magic_8_ball/run.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
+import io
```
Comment (abidlabs) [inline demo/magic_8_ball/run.py:1]: image

Sweet! I asked it what career I should choose

Btw the generation takes quite long -- can we cache some examples so that users can get an idea of what it sounds like?

File: demo/streaming_filter_unified/run.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    else:
+        return np.flipud(frame)
+
+
+css=""".my-group {max-width: 500px !important; max-height: 500px !important;}
+            .my-column {display: flex !important; justify-content: center !important; align-items: center !important};"""
+
+with gr.Blocks(css=css) as demo:
+    with gr.Tabs():
+        with gr.Tab("Code"):
```
Comment (abidlabs) [inline demo/streaming_filter_unified/run.py:40]: Do you intentionally have this code/demo tab? Why include the code like this?

File: demo/streaming_filter_unified/run.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
+import gradio as gr
```
Comment (abidlabs) [inline demo/streaming_filter_unified/run.py:1]: Nit: I think the "flip" is not the most compelling initial state for a demo showing the unified input/output image streaming. "edges" makes a bit more sense. Would even cooler if we could add a real object detection demo with YolovX

File: guides/04_additional-features/03_streaming-inputs.md [hunk 6]
```diff
@@ -0,0 +1,59 @@
+# Streaming inputs
+
+In the previous guide, we covered how to stream a sequence of outputs from an event handler. Gradio also allows you to stream images from a user's camera or audio chunks from their microphone **into** your event handler. This can be used to create real-time object detection apps or conversational chat applications with Gradio.
+
+Currently, the `gr.Image` and the `gr.Audio` components support input streaming via the `stream` event.
+Let's create the simplest streaming app possible, which simply returns the webcam stream unmodified.
+
+$demo_streaming_simple
```
Comment (abidlabs) [inline guides/04_additional-features/03_streaming-inputs.md:8]: I don't think we should be putting the code within the Gradio demo, it just makes things harder to maintain. We can perhaps change how $demo is rendered to include the code but that should happen at the website level

File: guides/04_additional-features/03_streaming-inputs.md [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+
+Try it out! The streaming event is triggered when the user starts recording. Under the hood, the webcam will take a photo every 0.1 seconds and send it to the server. The server will then return that image.
+
+There are two unique keyword arguments for the `stream` event:
+
+* `time_limit` - This is the amount of time the gradio server will spend processing the event. Media streams are naturally unbounded so it's important to set a time limit so that one user does not hog the Gradio queue. The time limit only counts the time spent processing the stream, not the time spent waiting in the queue. The orange bar displayed at the bottom of the input image represents the remaining time. When the time limit expires, the user will automatically rejoin the queue.
```
Comment (abidlabs) [inline guides/04_additional-features/03_streaming-inputs.md:14]: Perhap include the the types, e.g. 
```suggestion
* `time_limit (float)` - This is the amount of time the gradio server will spend processing the event. Media streams are naturally unbounded so it's important to set a time limit so that one user does not hog the Gradio queue. The time limit only counts the time spent processing the stream, not the time spent waiting in the queue. The orange bar displayed at the bottom of the input image represents the remaining time. When the time limit expires, the user will automatically rejoin the queue.
```

File: guides/09_other-tutorials/streaming-ai-generated-audio.md [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+
+A Magic 8 Ball is a toy that answers any question after you shake it. Our application will do the same but it will also speak its response!
+
+We won't cover all the implementation details in this blog post but the code is freely available on [Hugging Face Spaces](https://huggingface.co/spaces/gradio/magic-8-ball).
+
+## The Overview
+
+Just like the classic Magic 8 Ball, a user should ask it a question orally and then wait for a response. Under the hood, we'll use Whisper to transcribe the audio and then use an LLM to generate a magic-8-ball-style answer. Finally, we'll use Parler TTS to read the response aloud.
+
+## The UI
```
Comment (abidlabs) [inline guides/09_other-tutorials/streaming-ai-generated-audio.md:15]: Would be really cool if we could get the UI to "shake" with some custom JS to really mimic an 8 ball

**Review by abidlabs — APPROVED** [review]
Great writeups and fixes @freddyaboulton! Left a few suggestions but all minor

File: guides/09_other-tutorials/streaming-ai-generated-audio.md (hunk 8 above)
Comment (freddyaboulton) [reply guides/09_other-tutorials/streaming-ai-generated-audio.md:15]: Yes I actually thought about this but then ran into this issue! https://github.com/gradio-app/gradio/pull/9188

File: demo/magic_8_ball/requirements.txt (hunk 2 above)
Comment (freddyaboulton) [reply demo/magic_8_ball/requirements.txt:1]: We need this until we publish the rc I guess

File: js/spa/test/cancel_events.spec.ts (hunk 1 above)
Comment (freddyaboulton) [reply js/spa/test/cancel_events.spec.ts:10]: I thought it was related to the e2e changes but it's not!

File: demo/streaming_filter_unified/run.py (hunk 4 above)
Comment (freddyaboulton) [reply demo/streaming_filter_unified/run.py:40]: I think it helps associate the code with the demo and saves vertical scrolling space. Talking about building something more integrated into the website with @aliabd though. Context: https://huggingface.slack.com/archives/C03BHNJCBC4/p1724261743661269?thread_ts=1724256882.668899&cid=C03BHNJCBC4

File: guides/04_additional-features/03_streaming-inputs.md (hunk 6 above)
Comment (freddyaboulton) [reply guides/04_additional-features/03_streaming-inputs.md:8]: Yeap tracked here: https://huggingface.slack.com/archives/C03BHNJCBC4/p1724261743661269?thread_ts=1724256882.668899&cid=C03BHNJCBC4

Comment (freddyaboulton) [reply guides/04_additional-features/03_streaming-inputs.md:9]: Added back separate code

File: demo/streaming_filter_unified/run.py (hunk 5 above)
Comment (freddyaboulton) [reply demo/streaming_filter_unified/run.py:1]: Yes I think the current state is `flip`. A demo with yolo would be cool as well!

**Comment by freddyaboulton** [comment]
Thank you for the review @abidlabs !!
